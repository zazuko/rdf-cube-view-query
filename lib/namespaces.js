import $rdf from '@zazuko/env'

const cube = $rdf.namespace('https://cube.link/')
const view = $rdf.namespace('https://cube.link/view/')

const {
  rdf,
  schema,
  sh,
  xsd,
} = $rdf.ns

export { cube, rdf, schema, sh, view, xsd }
