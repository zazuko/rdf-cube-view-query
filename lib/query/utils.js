import rdf from '@zazuko/env'

export function contains(ptr, target) {
  if (!target) {
    return ptr.terms.length > 0
  }

  if (target.term) {
    target = target.term
  }

  if (target.ptr && target.ptr.term) {
    target = target.ptr.term
  }

  return ptr.terms.some(term => term.equals(target))
}

export function distinct(...ptrs) {
  const ptr = ptrs[0]
  const termSet = rdf.termSet()

  for (const ptr of ptrs) {
    for (const term of ptr.terms) {
      termSet.add(term)
    }
  }

  const terms = [...termSet]

  if (terms.length === 0) {
    return ptr
  }

  return rdf.clownface({
    dataset: ptr._context[0].dataset,
    term: terms,
    graph: ptr._context[0].graph,
  })
}

export function toPropertyPath(ptr) {
  const list = ptr.list()

  if (!list) {
    return ptr.term
  }

  return [...list].map(property => property.term)
}
