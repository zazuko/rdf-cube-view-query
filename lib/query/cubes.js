import rdf from '@zazuko/env'
import * as sparql from 'rdf-sparql-builder'
import * as ns from '../namespaces.js'

export function cubesQuery({ filters = [], graph = rdf.defaultGraph() } = {}) {
  const cube = rdf.variable('cube')

  const patterns = [[cube, ns.rdf.type, ns.cube.Cube]]

  for (const [index, filter] of filters.entries()) {
    const content = filter({ cube, index })

    for (const c of content) {
      patterns.push(c)
    }
  }

  return sparql.select([cube])
    .from(graph)
    .where(patterns)
    .toString()
}
