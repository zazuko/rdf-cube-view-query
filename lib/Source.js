import ParsingClient from 'sparql-http-client/ParsingClient.js'
import Cube from './Cube.js'
import Node from './Node.js'
import * as ns from './namespaces.js'
import { toTerm } from './utils.js'
import { cubesQuery } from './query/cubes.js'
import { viewListQuery } from './query/views.js'
import View from './View.js'
import { createPrefixedSparqlClient } from './PrefixedSparqlClient.js'

const DEFAULT_QUERY_PREFIX = '#pragma describe.strategy cbd'

export default class Source extends Node {
  constructor({
    parent,
    term,
    dataset,
    graph,
    endpointUrl,
    sourceGraph,
    user,
    password,
    queryOperation,
    queryPrefix,
    client = new ParsingClient({ endpointUrl, user, password }),
  }) {
    super({
      parent,
      term,
      dataset,
      graph,
    })

    this.ptr.addOut(ns.view.endpoint, toTerm(client.query.endpoint.endpointUrl))

    if (sourceGraph) {
      this.ptr.addOut(ns.view.graph, toTerm(sourceGraph))
    }

    this.user = user
    this.password = password
    this.queryOperation = queryOperation
    this.queryPrefix = queryPrefix || DEFAULT_QUERY_PREFIX
    this.client = createPrefixedSparqlClient({ client, queryPrefix: this.queryPrefix })
  }

  get endpoint() {
    return this.ptr.out(ns.view.endpoint).term
  }

  get graph() {
    return this.ptr.out(ns.view.graph).term
  }

  async cube(term) {
    const cube = new Cube({
      parent: this,
      term: toTerm(term),
      source: this,
    })

    await cube.init()

    // empty cube?
    if (cube.ptr.out().terms.length === 0) {
      return null
    }

    return cube
  }

  cubesQuery({ filters } = {}) {
    return cubesQuery({
      graph: this.graph,
      filters,
    })
  }

  async cubes({ filters, noShape = false } = {}) {
    const rows = await this.client.query.select(this.cubesQuery({ filters }))

    return Promise.all(rows.map(async row => {
      const cube = new Cube({
        parent: this,
        term: row.cube,
        source: this,
      })

      await cube.fetchCube()

      if (!noShape) {
        await cube.fetchShape()
      }

      return cube
    }))
  }

  viewListQuery({ filters } = {}) {
    return viewListQuery({
      graph: this.graph, filters,
    })
  }

  async views({ filters } = {}) {
    const rows = await this.client.query.select(this.viewListQuery({ filters }))

    return Promise.all(rows.map(async row => {
      const view = new View({
        parent: this, term: row.view, source: this,
      })

      await view.fetchView()

      return view
    }))
  }

  async view(term) {
    const view = new View({
      parent: this, term: toTerm(term), source: this,
    })

    await view.init()

    // empty view?
    if (view.ptr.out().terms.length === 0) {
      return null
    }

    return view
  }
}
