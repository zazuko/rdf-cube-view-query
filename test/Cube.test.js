import { strictEqual } from 'assert'
import rdfHandler from '@rdfjs/express-handler'
import withServer from 'express-as-promise/withServer.js'
import rdf from '@zazuko/env'
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { cubesQuery } from '../lib/query/cubes.js'
import Cube from '../lib/Cube.js'
import Source from '../lib/Source.js'
import { buildCube } from './support/buildCube.js'
import * as ns from './support/namespaces.js'
import { cleanQuery } from './support/utils.js'

describe('Cube', () => {
  chai.use(jestSnapshotPlugin())

  it('should be a constructor', () => {
    strictEqual(typeof Cube, 'function')
  })

  describe('dimensions', () => {
    it('should be an array property', () => {
      const cube = buildCube()

      strictEqual(Array.isArray(cube.dimensions), true)
    })

    it('should contain all dimensions of the cube without rdf:type', () => {
      const cube = buildCube({
        dimensions: [{
          path: ns.ex.propertyA,
        }, {
          path: ns.ex.propertyB,
        }],
      })

      const dimensions = cube.dimensions

      strictEqual(dimensions.length, 2)
      strictEqual(dimensions[0].path.equals(ns.ex.propertyA), true)
      strictEqual(dimensions[1].path.equals(ns.ex.propertyB), true)
    })
  })

  describe('fetchCube', () => {
    it('should be a method', () => {
      const cube = buildCube()

      strictEqual(typeof cube.fetchCube, 'function')
    })

    it('should use a DESCRIBE query to fetch the cube data', async () => {
      await withServer(async server => {
        let query = null

        server.app.get('/', rdfHandler(), (req, res) => {
          query = req.query.query

          res.dataset(rdf.dataset())
        })

        const cube = buildCube({ endpointUrl: await server.listen() })

        await cube.fetchCube()

        strictEqual(typeof query, 'string')
        strictEqual(query.includes('DESCRIBE'), true)
      })
    })

    it('should add the data from the endpoint to the dataset', async () => {
      await withServer(async server => {
        const quad = rdf.quad(ns.ex.subject, ns.ex.predicate, ns.ex.object)

        server.app.get('/', rdfHandler(), (req, res) => {
          res.dataset(rdf.dataset([quad]))
        })

        const cube = buildCube({ endpointUrl: await server.listen() })

        await cube.fetchCube()

        strictEqual(cube.ptr.dataset.size, 9)
        strictEqual(cube.ptr.dataset.has(quad), true)
      })
    })

    it('should add the data from the endpoint to the cleanup list', async () => {
      await withServer(async server => {
        const quad = rdf.quad(ns.ex.subject, ns.ex.predicate, ns.ex.object)

        server.app.get('/', rdfHandler(), (req, res) => {
          res.dataset(rdf.dataset([quad]))
        })

        const cube = buildCube({ endpointUrl: await server.listen() })

        await cube.fetchCube()

        strictEqual(cube.quads.length, 1)
        strictEqual(cube.quads.some(q => quad.equals(q)), true)
      })
    })
  })

  describe('fetchShape', () => {
    it('should be a method', () => {
      const cube = buildCube()

      strictEqual(typeof cube.fetchShape, 'function')
    })

    it('should use a DESCRIBE query to fetch the shape data', async () => {
      await withServer(async server => {
        let query = null

        server.app.get('/', rdfHandler(), (req, res) => {
          query = req.query.query

          res.dataset(rdf.dataset())
        })

        const cube = buildCube({ endpointUrl: await server.listen() })

        await cube.fetchShape()

        strictEqual(typeof query, 'string')
        strictEqual(query.includes('DESCRIBE'), true)
      })
    })

    it('should add the data from the endpoint to the dataset', async () => {
      await withServer(async server => {
        const quad = rdf.quad(ns.ex.subject, ns.ex.predicate, ns.ex.object)

        server.app.get('/', rdfHandler(), (req, res) => {
          res.dataset(rdf.dataset([quad]))
        })

        const cube = buildCube({ endpointUrl: await server.listen() })

        await cube.fetchShape()

        strictEqual(cube.ptr.dataset.size, 9)
        strictEqual(cube.ptr.dataset.has(quad), true)
      })
    })

    it('should add the data from the endpoint to the cleanup list', async () => {
      await withServer(async server => {
        const quad = rdf.quad(ns.ex.subject, ns.ex.predicate, ns.ex.object)

        server.app.get('/', rdfHandler(), (req, res) => {
          res.dataset(rdf.dataset([quad]))
        })

        const cube = buildCube({ endpointUrl: await server.listen() })

        await cube.fetchShape()

        strictEqual(cube.quads.length, 1)
        strictEqual(cube.quads.some(q => quad.equals(q)), true)
      })
    })
  })

  describe('filter', () => {
    describe('isPartOf', () => {
      it('should be a function', () => {
        strictEqual(typeof Cube.filter.isPartOf, 'function')
      })

      it('should create a patter in filter for schema:hasPart', async () => {
        const versionHistory = ns.ex.versionHistory
        const query = cubesQuery({
          filters: [Cube.filter.isPartOf(versionHistory)],
        })

        expect(cleanQuery(query)).toMatchSnapshot()
      })
    })

    describe('noExpires', () => {
      it('should be a function', () => {
        strictEqual(typeof Cube.filter.noExpires, 'function')
      })

      it('should create a not exists filter for schema:expires', async () => {
        const query = cubesQuery({
          filters: [Cube.filter.noExpires()],
        })

        expect(cleanQuery(query)).toMatchSnapshot()
      })
    })

    describe('noValidThrough', () => {
      it('should be a function', () => {
        strictEqual(typeof Cube.filter.noValidThrough, 'function')
      })

      it('should create a not exists filter for schema:validThrough', async () => {
        const query = cubesQuery({
          filters: [Cube.filter.noValidThrough()],
        })

        expect(cleanQuery(query)).toMatchSnapshot()
      })
    })

    describe('status', () => {
      it('should be a function', () => {
        strictEqual(typeof Cube.filter.status, 'function')
      })

      it('should create an in filter for the given status value', async () => {
        const query = cubesQuery({
          filters: [Cube.filter.status(ns.ex.status)],
        })

        expect(cleanQuery(query)).toMatchSnapshot()
      })

      it('should create an in filter for the given status values', async () => {
        const query = cubesQuery({
          filters: [Cube.filter.status([ns.ex.status1, ns.ex.status2])],
        })

        expect(cleanQuery(query)).toMatchSnapshot()
      })
    })

    describe('version', () => {
      it('should be a function', () => {
        strictEqual(typeof Cube.filter.version, 'function')
      })

      it('should create an eq filter for the given version', async () => {
        const query = cubesQuery({
          filters: [Cube.filter.version('2')],
        })

        expect(cleanQuery(query)).toMatchSnapshot()
      })

      const operations = ['eq', 'ne', 'lt', 'gt', 'lte', 'gte']

      for (const operation of operations) {
        describe(operation, () => {
          it('should be a function', () => {
            strictEqual(typeof Cube.filter.version[operation], 'function')
          })

          it(`should create an ${operation} filter for the given version`, async () => {
            const query = cubesQuery({
              filters: [Cube.filter.version[operation]('2')],
            })

            expect(cleanQuery(query)).toMatchSnapshot()
          })
        })
      }
    })
  })

  describe('in', () => {
    it('should be a method', () => {
      const source = new Source({ endpointUrl: ns.ex.endpoint })
      const cube = new Cube({ parent: source })

      strictEqual(typeof cube.in, 'function')
    })

    it('should use clownface to search for triples pointing to the cube', () => {
      const source = new Source({ endpointUrl: ns.ex.endpoint })
      const cube = new Cube({ parent: source })

      cube.ptr
        .addIn(ns.ex.predicate, ns.ex.up)
        .addOut(ns.ex.predicate, ns.ex.down)

      strictEqual(ns.ex.up.equals(cube.in(ns.ex.predicate).term), true)
    })
  })

  describe('out', () => {
    it('should be a method', () => {
      const source = new Source({ endpointUrl: ns.ex.endpoint })
      const cube = new Cube({ parent: source })

      strictEqual(typeof cube.out, 'function')
    })

    it('should use clownface to search for triples starting at the cube', () => {
      const source = new Source({ endpointUrl: ns.ex.endpoint })
      const cube = new Cube({ parent: source })

      cube.ptr
        .addIn(ns.ex.predicate, ns.ex.up)
        .addOut(ns.ex.predicate, ns.ex.down)

      strictEqual(ns.ex.down.equals(cube.out(ns.ex.predicate).term), true)
    })
  })
})
