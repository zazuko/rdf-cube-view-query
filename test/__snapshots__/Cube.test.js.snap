// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cube filter isPartOf should create a patter in filter for schema:hasPart 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>.
  <http://example.org/versionHistory> <http://schema.org/hasPart> ?cube.
}"
`;

exports[`Cube filter noExpires should create a not exists filter for schema:expires 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>.
  FILTER(NOT EXISTS { ?cube <http://schema.org/expires> ?v0. })
}"
`;

exports[`Cube filter noValidThrough should create a not exists filter for schema:validThrough 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>.
  FILTER(NOT EXISTS { ?cube <http://schema.org/validThrough> ?v0. })
}"
`;

exports[`Cube filter status should create an in filter for the given status value 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/creativeWorkStatus> ?v0.
  FILTER(?v0 IN(<http://example.org/status>))
}"
`;

exports[`Cube filter status should create an in filter for the given status values 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/creativeWorkStatus> ?v0.
  FILTER(?v0 IN(<http://example.org/status1>, <http://example.org/status2>))
}"
`;

exports[`Cube filter version eq should create an eq filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 = 2 )
}"
`;

exports[`Cube filter version gt should create an gt filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 > 2 )
}"
`;

exports[`Cube filter version gte should create an gte filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 >= 2 )
}"
`;

exports[`Cube filter version lt should create an lt filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 < 2 )
}"
`;

exports[`Cube filter version lte should create an lte filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 <= 2 )
}"
`;

exports[`Cube filter version ne should create an ne filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 != 2 )
}"
`;

exports[`Cube filter version should create an eq filter for the given version 1`] = `
"SELECT ?cube WHERE {
  ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>;
    <http://schema.org/version> ?v0.
  FILTER(?v0 = 2 )
}"
`;

exports[`Cube previewQuery creates a view query from the given cube 1`] = `
"SELECT ?dimension0 ?dimension1 WHERE {
  {
    SELECT ?source0 WHERE {
      <http://example.org/cube> <https://cube.link/observationSet> ?observationSet0.
      ?observationSet0 <https://cube.link/observation> ?source0.
    }
    LIMIT 10
  }
  ?source0 <http://example.org/propertyA> ?dimension0;
    <http://example.org/propertyB> ?dimension1.
}"
`;
